### ISSUE TITLE
Do I need to re-initiate weight & bias manually between folds when running cross validation ?

### ISSUE THREAD

**Issue by mark0935git** [skorch-dev/skorch#689]
I have been trying to learn skorch. I encountered an question is that whether I need to re-initiate weight & bias manually between folds when running cross validation (CV).

I was running the following script (just an example). I found that the model gave lower results and trained slower when I re-initiate weight & bias manually between folds in CV, comparing the model without "re-initiation".

How should we understand the differences? Do we have any explanations for the difference in model performance?

`from sklearn.model_selection import StratifiedKFold
from skorch import NeuralNetClassifier
from skorch.dataset import Dataset

class DNN4(nn.Module):
    def __init__(self,InD,OutD):
        super(DNN4,self).__init__()
        self.fc1 = nn.Linear(InD, 100)
        self.fc2 = nn.Linear(100, 400)
        self.fc3 = nn.Linear(400, 100)
        self.fc4 = nn.Linear(100, OutD)
        self.dropout = nn.Dropout(p=0.2)

    def forward(self, x):
        x = self.dropout(F.relu(self.fc1(x)))
        x = self.dropout(F.relu(self.fc2(x)))
        x = self.dropout(F.relu(self.fc3(x)))
        x = self.fc4(x)
        return x

def initialize_weights(self):
    for m in self.modules():
        print(m)
        if isinstance(m, nn.Linear):
            nn.init.xavier_uniform_(m.weight)
            nn.init.constant_(m.bias, 0)
    return

X = feature_array
Y = target_array
InModel=DNN4(InD=X.shape[1],OutD=2)

Fold = 1
SKFold = StratifiedKFold(n_splits=10, shuffle=True, random_state=123) #
for Train, Test in SKFold.split(X, Y):
    test_ds = Dataset(X[Test], y[Test])
    
    ###reinitiate weight & bias manually
    InModel.apply(initialize_weights)
    ###
        
    Model = NeuralNetClassifier(InModel, batch_size=256, max_epochs=100,
                              train_split=predefined_split(test_ds),
                              optimizer=torch.optim.Adam,
                              optimizer__lr=0.001,
                              optimizer__weight_decay=0.1,
                              criterion=nn.CrossEntropyLoss)
    
    
    Model.fit(X[Train], Y[Train])
    Fold+=1`

**Comment by mark0935git** [comment]
Have no cue why part of the code is recognized as code section. Sorry for the format.

**Comment by mark0935git** [comment]
@BenjaminBossan  Could you please help me with the question? Thanks a lot!

**Comment by BenjaminBossan** [comment]
You don't need to manually re-initialize weights (except if you set `warm_start=True`). However, what you should do differently is the following:

```python
model = NeuralNetClassifier(
    DNN4,
    module__InD=X.shape[1],
    module__OutD=2,
    ...
)
model.fit(...)
```

The difference to your code is that you don't pass an _already initialized_ module instance to `NeuralNetClassifier` but just the class instead. That way, you can be sure that with each fit call, a new instance of your module is created. Please test if this leads to the expected results.

**Comment by mark0935git** [comment]
> ```python
> X.shape[1]
> ```

**Comment by mark0935git** [comment]
Thanks for your instruction. I will try it out.
